utils/ensemble_tools.py

```diff
@@ -0,0 +1,158 @@
+""" A module to create input directories and encode
+configfurations for FabNESO
+"""
+
+import os
+import shutil
+import itertools
+
+
+def create_dir_tree(
+    sweep_path,
+    n_dirs,
+    destructive,
+    copy_dir,
+    edit_file,
+    parameter_to_scan,
+    scan_range,
+    outdir_prefix,
+):
+    """Create a directory tree in the sweep_path"""
+
+    copy_files = os.path.isdir(copy_dir)
+    if not copy_files:
+        print("No correct copy dir found, just " "creating the tree for now!")
+
+    # Make the base directory
+    if not make_directory(sweep_path, destructive):
+        return
+
+    # Set the initial value of the scanned parameter to the lower limit
+    para_val = scan_range[0]
+
+    for i in range(n_dirs):
+        newDir = "{0}/SWEEP/{2}{1}/".format(sweep_path, i, outdir_prefix)
+        # Make the directory
+        os.makedirs(newDir)
+        # If we're copying files, do so
+        if copy_files:
+            copy_dir_contents(newDir, copy_dir)
+        # Now we edit the parameter file for our
+        # template scan if we're doing that
+        if os.path.isfile(newDir + edit_file) and parameter_to_scan:
+            edit_parameter(newDir + edit_file, parameter_to_scan, para_val)
+        # iterate para_val
+        para_val += (scan_range[1] - scan_range[0]) / float(n_dirs - 1)
+
+
+def create_dict_sweep(
+    sweep_path, n_dirs, destructive, copy_dir, edit_file, parameter_dict
+):
+    """Use a dictionary with each parameter's high and low to create
+    a multi-dimensional sweep directory"""
+
+    print(
+        "{0} parameters found, requiring {1} divisions per parameter. "
+        "Creating {2} configurations...".format(
+            len(parameter_dict.keys()), n_dirs, n_dirs ** len(parameter_dict.keys())
+        )
+    )
```

Comment:
```suggestion
    print(
        f"{len(parameters_dict)} parameters found, "
        f"requiring {n_dirs} divisions per parameter. "
        f"Creating {n_dirs ** len(parameter_dict)} configurations..."
    )
```
[Formatted string literals (f-strings)](https://docs.python.org/3/tutorial/inputoutput.html#formatted-string-literals) are generally recommended over the `str.format` method in recent Python versions, as they're usually more readable and less likely to get bugs from putting arguments in wrong order.

The length of a `dict` is equivalent to the length of it's keys (or values) so I would go with the slightly shorter form.